Error reporting when building micro-execution modeling framework

**question 1**
Dockerfile：
```
FROM binaryanalysisplatform/bap@sha256:59eb9b54d386439d969fcb257b3fc5d3132e9b1bf230001bddcb8fe78fda565c

USER root

ENV DEBIAN_FRONTEND=noninteractive
RUN ln -fs /usr/share/zoneinfo/America/New_York /etc/localtime

RUN echo "deb http://dk.archive.ubuntu.com/ubuntu/ xenial main" >> /etc/apt/sources.list
RUN echo "deb http://dk.archive.ubuntu.com/ubuntu/ xenial universe" >> /etc/apt/sources.list

RUN apt update
RUN apt -y --no-install-recommends install build-essential vim python python3-pip tzdata pkg-config libgmp-dev zlib1g-dev

COPY build /build

ENV PATH=/build/bin:$PATH

WORKDIR /debug

RUN make

ENV LD_LIBRARY_PATH=/debug/upc.so

USER opam

RUN opam config exec -- opam install core core_kernel shexp yojson camlp4

COPY --chown=opam bap-plugin/abi/posix.h /home/opam/.opam/4.09/share/bap/api/c/posix.h
COPY --chown=opam bap-plugin/systems/core.asd /home/opam/.opam/4.09/share/bap/primus/systems/core.asd
COPY --chown=opam bap-plugin/site-lisp /home/opam/.opam/4.09/share/bap/primus/site-lisp

COPY --chown=opam . /upc

WORKDIR /upc/bap-plugin

# Set up the environment

ENV CAML_LD_LIBRARY_PATH=/home/opam/.opam/4.09/lib/stublibs:/home/opam/.opam/4.09/lib/ocaml/stublibs:/home/opam/.opam/4.09/lib/ocaml
ENV OCAML_TOPLEVEL_PATH=/home/opam/.opam/4.09/lib/toplevel
ENV OPAMYES=1
ENV OPAM_SWITCH_PREFIX=/home/opam/.opam/4.09
ENV PATH=/home/opam/.opam/4.09/bin:/home/opam/.opam/4.09/bin:/uIDS/uIDS/bin/:/build/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
ENV PKG_CONFIG_PATH=/home/opam/.opam/4.09/lib/pkgconfig

RUN opam config exec -- oasis setup -setup-update dynamic
RUN opam config exec -- make
RUN opam config exec -- make install
RUN opam config exec -- ./bin/build-plugin.sh

WORKDIR /upc/bap-plugin/entrypoint

RUN opam config exec -- make

WORKDIR /upc/bap-plugin/

RUN cp /upc/bap-plugin/entrypoint/upc /upc/bap-plugin/bin/upc

ENV PATH="/upc/bap-plugin/bin/:${PATH}"

RUN touch /tmp/srv.log

ENTRYPOINT ["/upc/bap-plugin/bin/upc"]
```
error:
```
#1 [internal] load build definition from Dockerfile
#1 transferring dockerfile: 2.10kB done
#1 DONE 0.0s

#2 [internal] load .dockerignore
#2 transferring context: 54B done
#2 DONE 0.0s

#3 [internal] load metadata for docker.io/binaryanalysisplatform/bap@sha256:59eb9b54d386439d969fcb257b3fc5d3132e9b1bf230001bddcb8fe78fda565c
#3 DONE 16.2s

#4 [ 1/29] FROM docker.io/binaryanalysisplatform/bap@sha256:59eb9b54d386439d969fcb257b3fc5d3132e9b1bf230001bddcb8fe78fda565c
#4 DONE 0.0s

#5 [internal] load build context
#5 transferring context: 8.65kB done
#5 DONE 0.0s

#6 [ 6/29] RUN apt -y --no-install-recommends install build-essential vim python python3-pip tzdata pkg-config libgmp-dev zlib1g-dev
#6 CACHED

#7 [ 4/29] RUN echo "deb http://dk.archive.ubuntu.com/ubuntu/ xenial universe" >> /etc/apt/sources.list
#7 CACHED

#8 [ 5/29] RUN apt update
#8 CACHED

#9 [ 2/29] RUN ln -fs /usr/share/zoneinfo/America/New_York /etc/localtime
#9 CACHED

#10 [ 3/29] RUN echo "deb http://dk.archive.ubuntu.com/ubuntu/ xenial main" >> /etc/apt/sources.list
#10 CACHED

#11 [ 7/29] COPY build /build
#11 ERROR: failed to calculate checksum of ref 808c5aaa-e191-42af-8cb2-fc5acd2def83::lopki7wvurbolgmari2zjkmc8: "/build": not found
------
 > [ 7/29] COPY build /build:
------
Dockerfile:14
--------------------
  12 |     RUN apt -y --no-install-recommends install build-essential vim python python3-pip tzdata pkg-config libgmp-dev zlib1g-dev
  13 |     
  14 | >>> COPY build /build
  15 |     
  16 |     ENV PATH=/build/bin:$PATH
--------------------
ERROR: failed to solve: failed to compute cache key: failed to calculate checksum of ref 808c5aaa-e191-42af-8cb2-fc5acd2def83::lopki7wvurbolgmari2zjkmc8: "/build": not found
```
additional:

Because the dockerfile contains the following code
`COPY --chown=opam bap-plugin/systems/core.asd /home/opam/.opam/4.09/share/bap/primus/systems/core.asd`
There is no core.asd file in the bap-plugin/systems folder, so I copied the core.asd in the bap-plugin/abi folder to the systems folder. I don’t know if this is correct.
After that I ran following command before hand:
`docker build -f docker/Dockerfile -t sysflowtelemetry/upolicycraft:latest .`

**question 2**
Because there is no` build` folder, I  created a `build` folder, but still got an error.
An error occurred when running `make` after `WORKDIR  debug` :
```
#1 [internal] load .dockerignore
#1 transferring context: 54B done
#1 DONE 0.0s

#2 [internal] load build definition from Dockerfile
#2 transferring dockerfile: 2.10kB done
#2 DONE 0.0s

#3 [internal] load metadata for docker.io/binaryanalysisplatform/bap@sha256:59eb9b54d386439d969fcb257b3fc5d3132e9b1bf230001bddcb8fe78fda565c
#3 DONE 16.2s

#4 [ 1/29] FROM docker.io/binaryanalysisplatform/bap@sha256:59eb9b54d386439d969fcb257b3fc5d3132e9b1bf230001bddcb8fe78fda565c
#4 DONE 0.0s

#5 [internal] load build context
#5 transferring context: 10.75kB done
#5 DONE 0.0s

#6 [ 7/29] COPY build /build
#6 CACHED

#7 [ 8/29] WORKDIR /debug
#7 CACHED

#8 [ 4/29] RUN echo "deb http://dk.archive.ubuntu.com/ubuntu/ xenial universe" >> /etc/apt/sources.list
#8 CACHED

#9 [ 5/29] RUN apt update
#9 CACHED

#10 [ 6/29] RUN apt -y --no-install-recommends install build-essential vim python python3-pip tzdata pkg-config libgmp-dev zlib1g-dev
#10 CACHED

#11 [ 2/29] RUN ln -fs /usr/share/zoneinfo/America/New_York /etc/localtime
#11 CACHED

#12 [ 3/29] RUN echo "deb http://dk.archive.ubuntu.com/ubuntu/ xenial main" >> /etc/apt/sources.list
#12 CACHED

#13 [ 9/29] RUN pwd && ls -la
#13 CACHED

#14 [10/29] RUN make
#0 0.138 make: *** No targets specified and no makefile found.  Stop.
#14 ERROR: process "/bin/sh -c make" did not complete successfully: exit code: 2
------
 > [10/29] RUN make:
#0 0.138 make: *** No targets specified and no makefile found.  Stop.
------
Dockerfile:20
--------------------
  18 |     WORKDIR /debug
  19 |     RUN pwd && ls -la
  20 | >>> RUN make
  21 |     
  22 |     ENV LD_LIBRARY_PATH=/debug/upc.so
--------------------
ERROR: failed to solve: process "/bin/sh -c make" did not complete successfully: exit code: 2
```

**question 3**
Based on the appeal error, I tried to make the following modifications
`COPY build /build` --> `COPY debug /debug`
I don't know if this is the right thing to do.I ran the following command：
`docker build -f docker/Dockerfile -t sysflowtelemetry/upolicycraft:latest --progress=plain .`

But it reported an error again：
```
#1 [internal] load .dockerignore
#1 transferring context: 54B done
#1 DONE 0.0s

#2 [internal] load build definition from Dockerfile
#2 transferring dockerfile: 2.06kB done
#2 DONE 0.0s

#3 [internal] load metadata for docker.io/binaryanalysisplatform/bap@sha256:59eb9b54d386439d969fcb257b3fc5d3132e9b1bf230001bddcb8fe78fda565c
#3 DONE 16.2s

#4 [ 1/29] FROM docker.io/binaryanalysisplatform/bap@sha256:59eb9b54d386439d969fcb257b3fc5d3132e9b1bf230001bddcb8fe78fda565c
#4 DONE 0.0s

#5 [internal] load build context
#5 transferring context: 10.71kB done
#5 DONE 0.0s

#6 [13/29] COPY --chown=opam bap-plugin/abi/posix.h /home/opam/.opam/4.09/share/bap/api/c/posix.h
#6 CACHED

#7 [10/29] RUN make
#7 CACHED

#8 [ 3/29] RUN echo "deb http://dk.archive.ubuntu.com/ubuntu/ xenial main" >> /etc/apt/sources.list
#8 CACHED

#9 [ 4/29] RUN echo "deb http://dk.archive.ubuntu.com/ubuntu/ xenial universe" >> /etc/apt/sources.list
#9 CACHED

#10 [17/29] WORKDIR /upc/bap-plugin
#10 CACHED

#11 [ 6/29] RUN apt -y --no-install-recommends install build-essential vim python python3-pip tzdata pkg-config libgmp-dev zlib1g-dev
#11 CACHED

#12 [ 5/29] RUN apt update
#12 CACHED

#13 [16/29] COPY --chown=opam . /upc
#13 CACHED

#14 [11/29] RUN opam config exec -- opam install core core_kernel shexp yojson camlp4
#14 CACHED

#15 [ 8/29] WORKDIR /debug
#15 CACHED

#16 [ 7/29] COPY debug /debug
#16 CACHED

#17 [ 2/29] RUN ln -fs /usr/share/zoneinfo/America/New_York /etc/localtime
#17 CACHED

#18 [14/29] COPY --chown=opam bap-plugin/systems/core.asd /home/opam/.opam/4.09/share/bap/primus/systems/core.asd
#18 CACHED

#19 [12/29] RUN ls -l
#19 CACHED

#20 [15/29] COPY --chown=opam bap-plugin/site-lisp /home/opam/.opam/4.09/share/bap/primus/site-lisp
#20 CACHED

#21 [ 9/29] RUN pwd && ls -la
#21 CACHED

#22 [18/29] RUN ls -l
#22 CACHED

#23 [19/29] RUN opam config exec -- oasis setup -setup-update dynamic
#23 0.589 E: Copyright must follow the convention '(C) 2008-2009 J.R. Hacker', here it is '(C) 2024'
#23 ERROR: process "/bin/sh -c opam config exec -- oasis setup -setup-update dynamic" did not complete successfully: exit code: 1
------
 > [19/29] RUN opam config exec -- oasis setup -setup-update dynamic:
#23 0.589 E: Copyright must follow the convention '(C) 2008-2009 J.R. Hacker', here it is '(C) 2024'
------
Dockerfile:45
--------------------
  43 |     ENV PKG_CONFIG_PATH=/home/opam/.opam/4.09/lib/pkgconfig
  44 |     
  45 | >>> RUN opam config exec -- oasis setup -setup-update dynamic
  46 |     RUN opam config exec -- make
  47 |     RUN opam config exec -- make install
--------------------
ERROR: failed to solve: process "/bin/sh -c opam config exec -- oasis setup -setup-update dynamic" did not complete successfully: exit code: 1
```

It reported a version error. Does this mean that the code to install oasis or oasis dependencies should be added before?

--- a/README.md
+++ b/README.md
@@ -83,7 +83,7 @@ can use to micro-execute the entrypoint.
 
 In order to build the micro execution modeling framework, you can do the following:
 
-    docker build -f docker/Dockerfile -t sysflowtelemetry/uPolicyCraft:latest .
+    docker build -f docker/Dockerfile -t sysflowtelemetry/upolicycraft:latest .
 
 Micro-executing a given entrypoint will naturally require the program to
 utilize external library dependencies.  Instead of micro-executing these

--- a/bap-plugin/_oasis
+++ b/bap-plugin/_oasis
@@ -4,11 +4,11 @@ Version:     0.1
 Synopsis:    uPolicyCraft
 Authors:     William Blair
 License:     Apache
-Copyrights:  (C) 2024
+Copyrights:  (C) 2024 William Blair
 Plugins:     META (0.4), DevFiles (0.4)
 BuildTools: ocamlbuild
 
-BuildDepends: ppx_jane, core_kernel
+BuildDepends: ppx_jane, core, core_kernel
 
 Library primus_upolicycraft_plugin
   Path: .

--- a/bap-plugin/entrypoint/dune
+++ b/bap-plugin/entrypoint/dune
@@ -1,4 +1,4 @@
 (executable
-  (name entrypoint)
+  (name upc)
   (libraries core core_unix core_unix.sys_unix core_unix.command_unix shexp.process yojson)
   (preprocess (pps ppx_jane)))

--- a/bap-plugin/entrypoint/dune-project
+++ b/bap-plugin/entrypoint/dune-project
@@ -19,7 +19,7 @@
  (name entrypoint)
  (synopsis "An entrypoint for running uPolicyCraft.")
  (description "A wrapper for executing uPolicyCraft on top of BAP.")
- (depends ocaml dune)
+ (depends ocaml core core_unix dune)
  (tags
   ("stateful system call specialization")))
 

--- a/bap-plugin/upolicycraft.ml
+++ b/bap-plugin/upolicycraft.ml
@@ -54,10 +54,11 @@ end
 
 module Param = struct
   open Config;;
+  (**
   manpage [
     `S "DESCRIPTION";
     `P "Derive a process model for a binary."
-  ]
+  ] *)
 
   let model = flag "model"
       ~doc:
@@ -2367,7 +2368,7 @@ module Monitor(Machine : Primus.Machine.S) = struct
           let opt = (v |> Value.to_word |> Bitvector.to_int) in
           match opt with
             Ok fd ->
-              let _ = Hashtbl.set s.files ~key:fd ~data:file in
+              let () = Hashtbl.set s.files ~key:fd ~data:file in
               {s with file_opened = None}
           | Error _ -> s)
 
@@ -2450,7 +2451,7 @@ module Monitor(Machine : Primus.Machine.S) = struct
     let model' = Yojson.Basic.pretty_to_string model in
     let jsonfile_path = Printf.sprintf "%s/%s.json" (get model_dir) (get model_name) in
     let oc = Out_channel.create jsonfile_path in
-    let _ = Printf.fprintf oc "%s" model' in
+    let () = Printf.fprintf oc "%s" model' in
     Out_channel.close oc
 
   (** Doesn't Core already provide this? *)
@@ -2577,7 +2578,7 @@ module Monitor(Machine : Primus.Machine.S) = struct
       Machine.args >>= fun args ->
       Machine.Global.get state >>= fun state' ->
       let {root_tid;nodes;graph;functions;written_vars;visited_blks} = state' in
-      let _ = Graphlib.to_dot (module EffectGraph)
+      let () = Graphlib.to_dot (module EffectGraph)
         ~node_attrs:(fun tid ->
           let context = try Hashtbl.find_exn functions tid
             with Not_found_s s ->
@@ -2594,7 +2595,7 @@ module Monitor(Machine : Primus.Machine.S) = struct
           [`Fontsize 8; `Label label;])
         ~channel:dotfile graph in
       let concrete_vars = (Var.Set.length written_vars) in
-      let _ = export_model root_tid nodes functions graph concrete_vars symbolic_vars (Blk.Set.length visited_blks) basic_blocks in
+      let () = export_model root_tid nodes functions graph concrete_vars symbolic_vars (Blk.Set.length visited_blks) basic_blocks in
       Machine.return()
     else
       Machine.Local.get state >>= fun state' ->
@@ -2694,8 +2695,7 @@ module Monitor(Machine : Primus.Machine.S) = struct
       let {callstack;symbols} = s in
       let name' = match addr_of_sub_name name with
                     None -> name
-                  | Some x ->
-                    x in
+                  | Some x -> Printf.sprintf "0x%x" x in
       {s with callstack=name'::callstack}
     )
 

--- a/docker/Dockerfile
+++ b/docker/Dockerfile
@@ -1,4 +1,4 @@
-FROM binaryanalysisplatform/bap@sha256:59eb9b54d386439d969fcb257b3fc5d3132e9b1bf230001bddcb8fe78fda565c
+FROM binaryanalysisplatform/bap@sha256:f614e5295ef45a78670d1298343520a614bebe8792d7e195920b19c3277d0069
 
 USER root
 
@@ -11,9 +11,7 @@ RUN echo "deb http://dk.archive.ubuntu.com/ubuntu/ xenial universe" >> /etc/apt/
 RUN apt update
 RUN apt -y --no-install-recommends install build-essential vim python python3-pip tzdata pkg-config libgmp-dev zlib1g-dev
 
-COPY build /build
-
-ENV PATH=/build/bin:$PATH
+COPY debug /debug
 
 WORKDIR /debug
 
@@ -26,7 +24,7 @@ USER opam
 RUN opam config exec -- opam install core core_kernel shexp yojson camlp4
 
 COPY --chown=opam bap-plugin/abi/posix.h /home/opam/.opam/4.09/share/bap/api/c/posix.h
-COPY --chown=opam bap-plugin/systems/core.asd /home/opam/.opam/4.09/share/bap/primus/systems/core.asd
+COPY --chown=opam bap-plugin/abi/core.asd /home/opam/.opam/4.09/share/bap/primus/systems/core.asd
 COPY --chown=opam bap-plugin/site-lisp /home/opam/.opam/4.09/share/bap/primus/site-lisp
 
 COPY --chown=opam . /upc
@@ -42,18 +40,30 @@ ENV OPAM_SWITCH_PREFIX=/home/opam/.opam/4.09
 ENV PATH=/home/opam/.opam/4.09/bin:/home/opam/.opam/4.09/bin:/uIDS/uIDS/bin/:/build/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
 ENV PKG_CONFIG_PATH=/home/opam/.opam/4.09/lib/pkgconfig
 
+# Build the BAP Plugin
+
 RUN opam config exec -- oasis setup -setup-update dynamic
 RUN opam config exec -- make
 RUN opam config exec -- make install
 RUN opam config exec -- ./bin/build-plugin.sh
 
+# Build the entrypoint for invoking BAP + the plugin
+
 WORKDIR /upc/bap-plugin/entrypoint
+
+# Look up how to install dependencies via dune.
+RUN opam install core_unix
 
 RUN opam config exec -- make
 
 WORKDIR /upc/bap-plugin/
 
-RUN cp /upc/bap-plugin/entrypoint/upc /upc/bap-plugin/bin/upc
+USER root
+
+RUN cp /upc/bap-plugin/entrypoint/_build/default/upc.exe  /upc/bap-plugin/bin/upc
+RUN chown opam:opam /upc/bap-plugin/bin/upc
+
+USER opam
 
 ENV PATH="/upc/bap-plugin/bin/:${PATH}"
 

